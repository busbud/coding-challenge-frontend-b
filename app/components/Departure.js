import React, { PropTypes } from 'react';
import classNames from 'classnames/bind';
import moment from 'moment';
import styles from '../styles/components/departure.css';
import { Row, Col } from 'react-flexbox-grid';

const cx = classNames.bind(styles);

const Departure = ({departure, destination, operator, origin}) => {
  console.log(operator);
  const departureTime = moment(departure.departure_time).format('h:mm a');
  const arrivalTime = moment(departure.arrival_time).format('h:mm a');

  console.log(departureTime);
  return (
    <div className={cx('departure-box')}>
      <Row>
        <Col xs={2}>
          <p>{ departureTime }</p>
          <i className="fa fa-arrow-down" aria-hidden="true"></i>
          <p>{ arrivalTime }</p>
        </Col>
        <Col xs={4}>
          <p>{destination.name}</p>
          <p>{origin.name}</p>

        </Col>
        <Col xs={3}>{operator.name}</Col>
        <Col xs={3}>Price</Col>
      </Row>
    </div>
  );
};

Departure.propTypes = {
  departure: PropTypes.object,
  operator: PropTypes.object,
  origin: PropTypes.object,
  destination: PropTypes.object
};

export default Departure;
