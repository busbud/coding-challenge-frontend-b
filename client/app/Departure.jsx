import React            from 'react'
import ReactTooltip     from 'react-tooltip'
import Icon             from './Icon.jsx'
import LoadingAnimation from './LoadingAnimation.jsx'

class Departure extends React.Component {

  constructor(props) {
    super(props);
    this.state = {};
    this.state.departure = this.props.data;
    this.formatPrice = this.formatPrice.bind(this);
  }
  
  formatPrice(number, currency) {
    return (number / 100).toLocaleString("en-US", {style:"currency", currency: this.props.config.currency});
  }

  render() {

    var dep = this.state.departure;
    
    if(dep) {

      let amenityIcons = Object.keys(dep.amenities)
        .filter(function(amenity) {
          return dep.amenities[amenity] === true;
        }.bind(this))
        .map(function(amenity, i) {
          return (
            <div 
              key={i} 
              data-tip={this.props.config.strings['AMENITY_'+amenity.toUpperCase()]} 
              data-for={'ToolTip'} 
              data-place='top'>
              <Icon src={'/images/icons/'+amenity+'.svg#Layer_1'}/>
            </div>
          );
        }.bind(this));

      return (
        <div className="departure" data-departureid={dep.id}>
          <div className="depatureInner">
        
            <div className="row">
            
              <div className="col-xs-6 col-sm-3">
                <div className="row">
                  <div className="col-xs-12 labelText">
                    {this.props.config.strings.PICKUP}
                  </div>
                  <div className="col-xs-12 time">
                    {dep.departure_display_time}
                  </div>
                  <div className="col-xs-12 location">
                    {dep.origin_location_name}
                  </div>
                </div>
              </div>
              
              <div className="col-xs-6 col-sm-3">
                <div className="row">
                  <div className="col-xs-12 labelText">
                    {this.props.config.strings.DROP_OFF}
                  </div>
                  <div className="col-xs-12 time">
                    {dep.arrival_display_time}
                  </div>
                  <div className="col-xs-12 location">
                    {dep.destination_location_name}
                  </div>
                </div>
              </div>
              
              <div className="col-xs-12 col-sm-3">
                <div className="row">
                  <div className="col-xs-12 labelText">
                    {dep.operator_display_name} {dep.amenities.display_name}
                  </div>
                </div>
                <div className="row">
                  <div className="col-xs-12 amenities">
                    {amenityIcons}
                  </div>
                </div>
              </div>

              <div className="col-xs-12 col-sm-3">
                <div className="row">
                  <div className="col-xs-6 col-sm-12 price">
                    {this.formatPrice(dep.prices.total)}<span className="currency">{this.props.config.currency} </span>
                  </div>
                  <div className="col-xs-6 col-sm-12">
                    <a className="button pull-right" href={dep.links.deeplink}>
                      {this.props.config.strings.BUY}
                    </a>
                  </div>
                </div>
              </div>
              
            </div>
          </div>
        </div>
      );      
    } else {

      return (
        <div className="departure loading">
          <LoadingAnimation />
        </div>
      )

    }
  }
}

export default Departure;