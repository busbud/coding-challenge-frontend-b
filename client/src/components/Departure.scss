@import "./variables.scss";

.departure {
  background-color: $lightGrey;
  color: $grey;
  width: 100%;
  box-shadow: 0 0 5px lighten($blue, 10%);
  border-radius: 5px;

  &__content {
    &__label {
      margin: 0.5rem 0;
      padding: 0.5rem 0;
      border-top: 2px solid $lightGrey;
      border-bottom: 2px solid $lightGrey;
      font-weight: bold;
      &:first-child {
        border-radius: 5px 5px 0 0;
        border-top: none;
      }
    }
    &__location {
      p {
        margin: 0.5em;
      }
    }
    &__duration {
      margin-top: 1em;
      padding-top: 0.5rem;
      border-top: 2px solid $lightGrey;
    }
  }

  &__reservation {
    button {
      padding: 0.5rem 2rem;
      width: 100%;
      margin-top: 1rem;
      font-size: 1.4rem;
      border: none;
      background-color: $blue;
      box-shadow: 0 0 5px lighten($blue, 10%);
      border-radius: 0 0 5px 5px;
      color: $lightGrey;
    }
  }

  &__footer,
  &__reservation__availability {
    display: none;
  }
}
