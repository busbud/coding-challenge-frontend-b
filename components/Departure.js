import PropTypes from "prop-types";
import format from "date-fns/format";

import { withTranslation } from "../i18n";
import { fonts } from "../theme";
import Button from "./Button";

const formatPrice = (price, currency) => {
  return `$${(price / 100).toFixed(2)} ${currency}`;
};

const formatTime = time => {
  return format(time, "HH:MM A");
};

const Departure = ({
  origin_city,
  origin_location,
  destination_city,
  destination_location,
  operator,
  departure_time,
  arrival_time,
  prices,
  t
}) => {
  return (
    <li>
      <div className="left">
        <div className="row">
          <p className="heading">{t("departs")}</p>
          <p>
            <strong>{formatTime(departure_time)}</strong> — {origin_city.name},{" "}
            {origin_location.name}
          </p>
        </div>
        <div className="row">
          <p className="heading">{t("arrives")}</p>
          <p>
            <strong>{formatTime(arrival_time)}</strong> —{" "}
            {destination_city.name}, {destination_location.name}
          </p>
        </div>
        <div className="row">
          <p className="heading">{t("operated-by")}</p> <p />
          <p>
            <strong>{operator.name}</strong>
          </p>
        </div>
      </div>
      <div className="right">
        <p className="price">{formatPrice(prices.total, prices.currency)}</p>
        <div className="button-wrapper">
          <Button>{t("buy-now")}</Button>
        </div>
      </div>
      <style jsx>{`
        li {
          background-color: white;
          border-radius: 8px;
          box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.24);
          display: grid;
          grid-template-columns: 1fr;
          grid-row-gap: 16px;
          grid-column-gap: 32px;
          padding: 20px;
        }
        @media (min-width: 640px) {
          li {
            grid-template-columns: 1fr 220px;
          }
        }
        .row {
          display: flex;
          flex-wrap: wrap;
        }
        .row + .row {
          margin-top: 16px;
        }
        @media (min-width: 800px) {
          .row {
            flex-wrap: nowrap;
          }
        }
        .heading {
          flex-basis: 100%;
          min-width: 150px;
          padding-top: 4px;
        }
        @media (min-width: 800px) {
          .heading {
            flex-basis: auto;
          }
        }
        p {
          margin: 0;
        }
        strong {
          font-family: ${fonts.slab};
          font-size: 1.25rem;
        }
        @media (min-width: 640px) {
          .right {
            align-items: flex-end;
            display: flex;
            flex-direction: column;
            justify-content: space-between;
          }
        }
        .price {
          font-family: ${fonts.slab};
          font-size: 2rem;
          margin: 0;
        }
        .button-wrapper {
          margin-top: 12px;
        }
      `}</style>
    </li>
  );
};

Departure.propTypes = {
  origin_city: PropTypes.object,
  origin_location: PropTypes.object,
  destination_city: PropTypes.object,
  destination_location: PropTypes.object,
  operator: PropTypes.object,
  departure_time: PropTypes.string,
  arrival_time: PropTypes.string,
  prices: PropTypes.object,
  t: PropTypes.func.isRequired
};

export default withTranslation("common")(Departure);
