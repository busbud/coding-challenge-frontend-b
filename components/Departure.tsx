import React from 'react';
import Image from 'next/image';
import Card from '@mui/material/Card';
import CardHeader from '@mui/material/CardHeader';
import CardContent from '@mui/material/CardContent';
import Avatar from '@mui/material/Avatar';
import dayjs from 'dayjs';
import PurchaseButton from './PurchaseButton';

import { DeparturesFullEntity } from '../interfaces/response';

const dateFormat = 'hh:mm a';

const Departure = ({ departure }: { departure: DeparturesFullEntity}) => {
    return (
        <Card sx={{ width: '80%', m: 3, minWidth: '300px' }}>
            <CardHeader
                avatar={
                    <Avatar aria-label={departure.operator.name}>
                        <Image src={departure.operator.logo_url} layout='fill' />
                    </Avatar>
                }
                action={
                    <PurchaseButton price={departure.prices.total} link={departure.links.deeplink} />
                }
                title={`${departure.originLocation.cityName} - ${departure.originLocation.name}`}
                subheader={`${departure.destinationLocation.cityName} - ${departure.destinationLocation.name}`}
            />
            <CardContent>
                Departure: {dayjs(departure.departure_time).format(dateFormat)}
                <br/>
                Arrival: {dayjs(departure.arrival_time).format(dateFormat)}
            </CardContent>
        </Card>
    )
};

export default Departure;