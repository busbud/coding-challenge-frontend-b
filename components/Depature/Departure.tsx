import { IXDeparture } from '../../types'
import { useAppContext } from '../../store'
import { getTime } from '../../utils/dateHelpers'
import { formatToCurrency } from '../../utils/currencyHelpers'

interface IProps {
  departure: IXDeparture
}

export default function Depature({ departure }: IProps) {
  const [{ locations, cities }] = useAppContext()

  const departureLocation = locations[departure.origin_location_id]
  const departureTime = getTime(departure.departure_time)
  const departureCity = cities[departureLocation.city_id]

  const arrivalLocation = locations[departure.destination_location_id]
  const arrivalTime = getTime(departure.arrival_time)
  const arrivalCity = cities[arrivalLocation.city_id]

  const amount = formatToCurrency(departure.prices.total, departure.prices.currency, true)

  return (
    <div className="departure-container">
      <div className="departure-info">
        <p className="time">{departureTime}</p>
        <p className="city">{departureCity.name}</p>
        <p className="location">{departureLocation.name}</p>
      </div>

      <div className="route-sign">
        <span className="arrow"></span>
      </div>

      <div className="arrival-info">
        <p className="time">{arrivalTime}</p>
        <p className="city">{arrivalCity.name}</p>
        <p className="location">{arrivalLocation.name}</p>
      </div>

      <div className="price-info">
        <p className="price">{amount}</p>
      </div>
    </div>
  )
}