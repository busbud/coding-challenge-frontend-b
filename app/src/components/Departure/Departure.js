import React from "react";

import "./Departure.scss";

const Departure = () => <div id="departure">Departure</div>;

export default Departure;
