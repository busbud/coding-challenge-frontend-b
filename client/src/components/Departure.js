import React from "react";
import LocalizedStrings from "react-localization";
import "./Departure.scss";

let strings = new LocalizedStrings({
  en: {
    from: "From",
    to: "To",
    when: "When",
    book: "Book",
    duration: "Duration: ",
    complete: "Complete",
    seats: "Available seats:"
  },
  fr: {
    from: "Depuis",
    to: "Vers",
    when: "Date",
    book: "Réserver",
    duration: "Durée : ",
    complete: "Complet",
    seats: "Sièges disponibles :"
  }
});

const Departure = ({
  arrival_time,
  available_seats,
  complete,
  departure_time,
  duration,
  prices,
  seatClass,
  from,
  to,
  lang
}) => {
  strings.setLanguage(lang);
  return (
    <div className={`departure ${complete ? "departure--complete" : ""}`}>
      <div className="departure__content">
        <p className="departure__content__label">{strings.from}</p>
        <div className="departure__content__location">
          <p>{from}</p>
          <p>{new Date(departure_time).toLocaleString(lang)}</p>
        </div>
        <p className="departure__content__arrow">&rarr;</p>
        <p className="departure__content__label">{strings.to}</p>
        <div className="departure__content__location">
          <p>{to}</p>
          <p>{new Date(arrival_time).toLocaleString(lang)}</p>
        </div>
        <div className="departure__content__duration">
          <span>{strings.duration}</span>
          {duration}
        </div>
      </div>
      <div className="departure__reservation">
        <div className="departure__reservation__availability">
          <div>{seatClass}</div>
          <p>
            {complete
              ? strings.complete
              : `${strings.seats} ${available_seats}`}
          </p>
        </div>
        <div className="departure__reservation__price">
          {prices.total} {prices.currency}
        </div>
        <button>{strings.book}</button>
      </div>
    </div>
  );
};

export default Departure;
