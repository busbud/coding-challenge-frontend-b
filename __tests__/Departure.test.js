import React from "react";
import { mount, shallow } from "enzyme";
import toJSON from "enzyme-to-json";

import Departure from "../components/Departure";

import { departures } from "../lib/testData";

const {
  origin_city,
  origin_location,
  destination_city,
  destination_location,
  operator,
  departure_time,
  arrival_time,
  prices
} = departures[0];

jest.mock("react-i18next", () => ({
  withTranslation: () => Component => {
    Component.defaultProps = {
      ...Component.defaultProps,
      i18n: { language: "en" },
      t: key => key
    };
    return Component;
  }
}));

describe("<Departure />", () => {
  it("renders and matches the snapshot", () => {
    const wrapper = shallow(
      <Departure
        origin_location={origin_location}
        origin_city={origin_city}
        destination_location={destination_location}
        destination_city={destination_city}
        operator={operator}
        departure_time={departure_time}
        arrival_time={arrival_time}
        prices={prices}
      />
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("displays the departure info", () => {
    const wrapper = mount(
      <Departure
        origin_location={origin_location}
        origin_city={origin_city}
        destination_location={destination_location}
        destination_city={destination_city}
        operator={operator}
        departure_time={departure_time}
        arrival_time={arrival_time}
        prices={prices}
      />
    );

    const departureTime = wrapper.find(".departure-time");
    expect(departureTime.text()).toBe("01:30 AM");
    const arrivalTime = wrapper.find(".arrival-time");
    expect(arrivalTime.text()).toBe("08:30 AM");
    const price = wrapper.find(".price");
    expect(price.text()).toBe("$52.00 CAD");
  });

  it("displays the origin city and location info", () => {
    const wrapper = mount(
      <Departure
        origin_location={origin_location}
        origin_city={origin_city}
        destination_location={destination_location}
        destination_city={destination_city}
        operator={operator}
        departure_time={departure_time}
        arrival_time={arrival_time}
        prices={prices}
      />
    );

    const originCityName = wrapper.find(".origin-city-name");
    expect(originCityName.text()).toBe("New York");
    const originLocationName = wrapper.find(".origin-location-name");
    expect(originLocationName.text()).toBe("Port Authority");
  });

  it("displays the destination city and location info", () => {
    const wrapper = mount(
      <Departure
        origin_location={origin_location}
        origin_city={origin_city}
        destination_location={destination_location}
        destination_city={destination_city}
        operator={operator}
        departure_time={departure_time}
        arrival_time={arrival_time}
        prices={prices}
      />
    );

    const destinationCityName = wrapper.find(".destination-city-name");
    expect(destinationCityName.text()).toBe("Montreal");
    const destinationLocationName = wrapper.find(".destination-location-name");
    expect(destinationLocationName.text()).toBe("Gare d'autocars");
  });

  it("displays the operator info", () => {
    const wrapper = mount(
      <Departure
        origin_location={origin_location}
        origin_city={origin_city}
        destination_location={destination_location}
        destination_city={destination_city}
        operator={operator}
        departure_time={departure_time}
        arrival_time={arrival_time}
        prices={prices}
      />
    );

    const operatorName = wrapper.find(".operator-name");
    expect(operatorName.text()).toBe("Greyhound");
  });
});
