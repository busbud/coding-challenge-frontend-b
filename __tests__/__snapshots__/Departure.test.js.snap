// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Departure /> renders and matches the snapshot 1`] = `
<li
  className="jsx-141613734"
>
  <div
    className="jsx-141613734 left"
  >
    <div
      className="jsx-141613734 row"
    >
      <p
        className="jsx-141613734 heading"
      >
        departs
      </p>
      <p
        className="jsx-141613734"
      >
        <strong
          className="jsx-141613734 departure-time"
        >
          01:30 AM
        </strong>
         
        — 
        <span
          className="jsx-141613734 origin-city-name"
        >
          New York
        </span>
        ,
         
        <span
          className="jsx-141613734 origin-location-name"
        >
          Port Authority
        </span>
      </p>
    </div>
    <div
      className="jsx-141613734 row"
    >
      <p
        className="jsx-141613734 heading"
      >
        arrives
      </p>
      <p
        className="jsx-141613734"
      >
        <strong
          className="jsx-141613734 arrival-time"
        >
          08:30 AM
        </strong>
         
        —
         
        <span
          className="jsx-141613734 destination-city-name"
        >
          Montreal
        </span>
        ,
         
        <span
          className="jsx-141613734 destination-location-name"
        >
          Gare d'autocars
        </span>
      </p>
    </div>
    <div
      className="jsx-141613734 row"
    >
      <p
        className="jsx-141613734 heading"
      >
        operated-by
      </p>
       
      <p
        className="jsx-141613734"
      />
      <p
        className="jsx-141613734"
      >
        <strong
          className="jsx-141613734 operator-name"
        >
          Greyhound
        </strong>
      </p>
    </div>
  </div>
  <div
    className="jsx-141613734 right"
  >
    <p
      className="jsx-141613734 price"
    >
      $52.00 CAD
    </p>
    <div
      className="jsx-141613734 button-wrapper"
    >
      <Button>
        buy-now
      </Button>
    </div>
  </div>
  <JSXStyle
    dynamic={
      Array [
        "'Changa One', cursive",
        "'Changa One', cursive",
      ]
    }
    id="3131558381"
  >
    li.__jsx-style-dynamic-selector{background-color:white;border-radius:8px;box-shadow:0 1px 3px 0 rgba(0,0,0,0.24);display:grid;grid-template-columns:1fr;grid-row-gap:16px;grid-column-gap:32px;padding:20px;}@media (min-width:640px){li.__jsx-style-dynamic-selector{grid-template-columns:1fr 220px;}}.row.__jsx-style-dynamic-selector{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-wrap:wrap;-ms-flex-wrap:wrap;flex-wrap:wrap;}.row.__jsx-style-dynamic-selector+.row.__jsx-style-dynamic-selector{margin-top:16px;}@media (min-width:800px){.row.__jsx-style-dynamic-selector{-webkit-flex-wrap:nowrap;-ms-flex-wrap:nowrap;flex-wrap:nowrap;}}.heading.__jsx-style-dynamic-selector{-webkit-flex-basis:100%;-ms-flex-preferred-size:100%;flex-basis:100%;min-width:150px;padding-top:4px;}@media (min-width:800px){.heading.__jsx-style-dynamic-selector{-webkit-flex-basis:auto;-ms-flex-preferred-size:auto;flex-basis:auto;}}p.__jsx-style-dynamic-selector{margin:0;}strong.__jsx-style-dynamic-selector{font-family:'Changa One', cursive;font-size:1.25rem;}@media (min-width:640px){.right.__jsx-style-dynamic-selector{-webkit-align-items:flex-end;-webkit-box-align:flex-end;-ms-flex-align:flex-end;align-items:flex-end;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:column;-ms-flex-direction:column;flex-direction:column;-webkit-box-pack:justify;-webkit-justify-content:space-between;-ms-flex-pack:justify;justify-content:space-between;}}.price.__jsx-style-dynamic-selector{font-family:'Changa One', cursive;font-size:2rem;margin:0;}.button-wrapper.__jsx-style-dynamic-selector{margin-top:12px;}
  </JSXStyle>
</li>
`;
